dali/test/dali_operator_test.h

```diff
@@ -0,0 +1,165 @@
+#ifndef DALI_DALI_OPERATOR_TEST_H
+#define DALI_DALI_OPERATOR_TEST_H
+
+#include <gtest/gtest.h>
+#include "dali/pipeline/pipeline.h"
+#include "dali/pipeline/data/backend.h"
+
+namespace dali {
+
+namespace util {
+
+template<typename T>
+struct is_tuple : std::false_type {
+};
+
+template<typename... T>
+struct is_tuple<std::tuple<T...>> : std::true_type {
+};
+
+
+/// This is a helper function, because there's no `if constexpr` in C++11
+template<typename Backend>
+std::string BackendStringName() {
+  DALI_FAIL("Selected Backend is not supported");
+}
+
+
+template<>
+std::string BackendStringName<CPUBackend>() {
+  return "cpu";
+}
+
+
+template<>
+std::string BackendStringName<GPUBackend>() {
+  return "gpu";
+}
+
+}  // namespace util
+
+template<typename Inputs, typename Outputs>
+class DaliOperatorTest : public ::testing::Test {
+//    static_assert(util::is_tuple<Inputs>::value, "Inputs has to be either a tuple or a void");
+//    static_assert(util::is_tuple<Outputs>::value, "Outputs has to be a tuple");
+
+public:
+  using Arguments =  std::map<std::string, float>; // TODO boost::any
+
+  template<typename Backend>
+  void RunTest(Arguments operator_arguments, Outputs anticipated_outputs) {
+    auto op_spec = CreateOpSpec<Backend>(operator_name_, operator_arguments);
+    AddOperatorToPipeline(*pipeline_, op_spec);
+    AddInputsToPipeline<Backend>(*pipeline_, "input");
+    outputs_ = RunPipeline<Backend>(*pipeline_, op_spec);
+    ASSERT_TRUE(Verify(outputs_, anticipated_outputs));
+  }
+
+
+private:
+  virtual Inputs SetInputs() const = 0;  // TODO std::vector
+
+  virtual std::string SetOperator() const = 0; // TODO std::vector
+
+  virtual bool Verify(Outputs outputs, Outputs anticipated_outputs) const = 0;
```

Comment:
I think that Verify function should be implemented by this class and should not be overridable by user. User should only provide the expected output Tensors.

Reply:
The idea is not to have this class do everything. As discussed, the main goals are:
    Launching tests with fully customizable input data
    Transparently and consistently manage the creation, execution, and ownership of the underlying pipeline
    Easily access the outputs for verification

In that context, you should be able to create a fixture by inheriting from this class and implement your custom verification.

Reply:
The idea is not to have this class do everything. As discussed, the main goals are:
Launching tests with fully customizable input data
Transparently and consistently manage the creation, execution, and ownership of the underlying pipeline
Easily access the outputs for verification

In that context, you should be able to create a fixture by inheriting from this class and implement your custom verification.

Reply:
On top of what @pribalta pointed out, let's imagine such 2 use cases:
1. We'd like to test a Downsample operator, which uses e.g. nn interpolation. In this case providing output tensor is not enough, since we have to compare taking floating point precision into acount.
1. On the other hand, we'd like to test the same operator without any interpolation, there could be nice to test with precise values

Hence, we already have 2 verification methods, depending on use case

Reply:
So we should probably have some subclass/helper function that allows to compare the output tensors with expected tensors, as I guess, most of the test will just default to that.

That way that frameworks usually supply user with variety of asserts that compare two results, I would imagine we should have predefined ones for Tensors. They could allow for parametrization how exactly the comparison is done (even a custom lambda for comparing individual elements between tensors) to handle the case that @szalpal described.

Reply:
Sure, I agree that we should have that. Only not as a subclass, since there's no benefit over using external function for that. Nevertheless, in my opinion, it shouldn't be a part of this PR. Just whenever somebody needs any kind of verification function, he can implement it

Reply:
The point here, that I want to make is that most people will probably use the same way of comparing expected tensor to that one obtained from test. Having it present in test framework reduces the burden that is placed on test implementer. 

I'm not forcing the subclass idea, but as you have to derive from google's Test class, you could have a hierarchy of more or less automatized tests, and choose which one you want to use/extend.

Reply:
So I would like to see something like: 
```
DaliOpTest <-- DaliOpTestWithVerification <--
      ^                                      |
      |--------DaliOpTestWithFs   <------- DaliOpTestWithFSAndVerification
```
Don't pay attention to naming. Of course, I do not force that approach, I probably would be better to have this as helper functions, mine idea was coming from that Test is a class in Gtest.

Reply:
I think you are right in terms, that we need a component for testing FileSystem-based data. However, I don't feel like it needs to be inside inheritance procedures. I'd keep it in a composition manner, so we don't need to use multiple inheritance. I find multiple inheritance something that is not supposed to be used :)

Reply:
Just to recap:
I don't particularly care about the interface (I explained why mine shows interface), I am perfectly fine if it would be possible to get the same functionality simply with free functions.

Still I would like to get:
- test that can automatically check if operator for given tensors returns identical tensors to those provided by reference, and user is not bored to death with writing comparators themselves.
- have functionality for input and reference outputs in FS.

Please do not focus on just the second.

Reply:
I sill don't understand, what you have in mind here. Could you please give me an example, how would you see it look like?
```
class SomeOperatorTest : DaliOperatorTest<SomeInputType, SomeOutputType> {
  bool Verify(SomeOutputType output, SomeOutputType anticipated_output) {
    ???
  }
};
```

Reply:
So I am starting to think, that not exposing the actual result of the operation, albeit useful from the separation from Tensor and Dali POV, limits the ability to truly verify the result of the operation. What if I want to check if operator returns a Tensor with specific shape, type, maybe padding, and memory backend? Should I hack it and provide custom conversion from Tensor[List] to SomeOutputType that will do the checking?

If we were to have access to output Tensor, than we will be able to automate implementation of verify, just by doing:
```
bool Verify(Tensor output, SomeOutputType anticipated_output) const override {
    testing::CompareTensors(output, testing::ToTensor(aticipated_output));
  }
```
or 
```
bool Verify(Tensor output, Tensor anticipated_output) const override {
    testing::CompareTensors(output, anticipated_output);
  }
```
or even have the above defined in
```
class DaliOperatorCompareTensorsTest<SomeInputType> : DaliOperatorTest<SomeInputType, Tensor> {
bool Verify(Tensor output, Tensor anticipated_output) const override {
    testing::CompareTensors(output, anticipated_output);
  }
}
```

Reply:
Having `Tensor` here breaks 3 of 4 applicable SOLID principles:
1. SRP - changes in Tensor affect `DaliOperatorTest`
2. OCP - code is not closed for changes, cause in case Tensor changes, code needs to change
3. LSP - that's not broken
5. DIP - cause DaliOperatorTest doesn't (in case of Tesnor) depend on abstraction, but on specific class, which is subject for changes.
I'm not saying, they can't be broken (e.g. using `std::string` breaks DIP, but I don't really mind, since `std::string` is really reliable class, that won't change in next 50 or so years). But there have to be reasons. Could you provide reasons for breaking them?

> Should I hack it and provide custom conversion from Tensor[List] to SomeOutputType that will do the checking?

Not *hack*, but *extend*. It's enormous difference.

About the latter example, that's just baaad. It can be implemented using composition, so why use inheritance? On top of that, imagine, that you would like to include also `MagicTensor`, `MyTensor` and `NotPreciselyATensor` into comparison. This way you would need to implement Cartesian products of all these types...

Reply:
> Having Tensor here breaks 3 of 4 applicable SOLID principles

As I said bellow in review summary, you still get a Tensor-like data here (contiguous buffer + shape). You can probably use a simple abstraction over it, and treat it as a common output data for tests, but still you have to provide the tests for the conversion.

In current case, you will have to write the test and the test for your FromTensorList and ToTensorList functions (that is add tests to your tests).

> Not hack, but extend. It's enormous difference.

It's not a difference if I have to use an API for the purpose that it was not intended to be used.

Tell me, if I write a test for resize or crop, how can I easily check if the output Tensors are of the intended size?

> On top of that, imagine, that you would like to include also MagicTensor, MyTensor and NotPreciselyATensor into comparison. This way you would need to implement Cartesian products of all these types...

So I imagine, that you could use the more generic test class then, and I don't see the problem here.


At this point it would be probably nice if somebody else also shared theirs opinion on the tests, as I think we are stuck.

Reply:
Also another thing:
maybe it should be done from both sides the same way (in regard to separation), at least in the simpler cases:
you provide `Input` and `Output ant_out`, which are converted to TensorLists (by the test framework, the conversion routines are tested as well).

The test is run with input data (obtained TensorList), and the output of the test is compared to `ToTensorList(ant_out)` using one of the chosen comparers.